src/llmq/quorums_blockprocessor.cpp

```diff
@@ -0,0 +1,443 @@
+                LogPrintf("CQuorumBlockProcessor::%s -- block %s not in active chain, peer=%d\n", __func__,
+                          qc.quorumHash.ToString(), pfrom->id);
+                // same, can't punish
+                return;
+            }
+            int quorumHeight = pquorumIndex->nHeight - (pquorumIndex->nHeight % params.dkgInterval);
+            if (quorumHeight != pquorumIndex->nHeight) {
+                LogPrintf("CQuorumBlockProcessor::%s -- block %s is not the first block in the DKG interval, peer=%d\n", __func__,
+                          qc.quorumHash.ToString(), pfrom->id);
+                Misbehaving(pfrom->id, 100);
+                return;
+            }
+        }
+
+        {
+            // Check if we already got a better one locally
+            // We do this before verifying the commitment to avoid DoS
+            LOCK(minableCommitmentsCs);
+            auto k = std::make_pair(type, qc.quorumHash);
+            auto it = minableCommitmentsByQuorum.find(k);
+            if (it != minableCommitmentsByQuorum.end()) {
+                auto jt = minableCommitments.find(it->second);
+                if (jt != minableCommitments.end()) {
+                    if (jt->second.CountSigners() <= qc.CountSigners()) {
+                        return;
+                    }
+                }
+            }
+        }
+
+        auto members = CLLMQUtils::GetAllQuorumMembers(type, qc.quorumHash);
+
+        if (!qc.Verify(members)) {
+            LOCK(cs_main);
+            LogPrintf("CQuorumBlockProcessor::%s -- commitment for quorum %s:%d is not valid, peer=%d\n", __func__,
+                      qc.quorumHash.ToString(), qc.llmqType, pfrom->id);
+            Misbehaving(pfrom->id, 100);
+            return;
+        }
+
+        LogPrintf("CQuorumBlockProcessor::%s -- received commitment for quorum %s:%d, validMembers=%d, signers=%d, peer=%d\n", __func__,
+                  qc.quorumHash.ToString(), qc.llmqType, qc.CountValidMembers(), qc.CountSigners(), pfrom->id);
+
+        AddMinableCommitment(qc);
+    }
+}
+
+bool CQuorumBlockProcessor::ProcessBlock(const CBlock& block, const CBlockIndex* pindexPrev, CValidationState& state)
+{
+    AssertLockHeld(cs_main);
+
+    bool fDIP0003Active = VersionBitsState(pindexPrev, Params().GetConsensus(), Consensus::DEPLOYMENT_DIP0003, versionbitscache) == THRESHOLD_ACTIVE;
+    if (!fDIP0003Active) {
+        return true;
+    }
+
+    int nHeight = pindexPrev->nHeight + 1;
+
+    std::map<Consensus::LLMQType, CFinalCommitment> qcs;
+    if (!GetCommitmentsFromBlock(block, pindexPrev, qcs, state)) {
+        return false;
+    }
+
+    for (const auto& p : Params().GetConsensus().llmqs) {
+        auto type = p.first;
+
+        uint256 quorumHash = GetQuorumBlockHash(type, pindexPrev);
+
+        if (!quorumHash.IsNull() && IsMiningPhase(type, nHeight)) {
+            if (HasMinedCommitment(type, quorumHash)) {
+                if (qcs.count(type)) {
```

Comment:
All these nested `if` / `else` s start to worry me a bit. 😅  Is there a possibility of squashing these inner `ifs` into a single joined by `&&`, or is it expected that `HasMinedCommitment(type, quorumHash)` might happen w/o `qcs.count(type)`?

Maybe I'm just too paranoid about nested if/elses, but I've seen a lot of bugs sneak into these in my experience.

Reply:
Hmm yeah, the code was unnecessarily nested. I refactored it and hope it's easier to read now